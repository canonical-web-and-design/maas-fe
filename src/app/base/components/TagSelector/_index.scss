@mixin TagSelector {
  $dropdown-max-height: 20rem;

  .tag-selector {
    position: relative;
  }

  .tag-selector .p-form-validation__message {
    margin-top: 0;
  }

  .tag-selector__input {
    position: relative;
    cursor: pointer;

    &.tags-selected {
      border-top: 0;
      box-shadow: none;
      top: -#{$border-radius};
    }

    &[disabled],
    &[disabled="disabled"] {
      opacity: 1;
    }
  }

  .tag-selector__dropdown {
    @extend %vf-bg--x-light;
    @extend %vf-has-box-shadow;
    left: 0;
    position: absolute;
    right: 0;
    top: calc(100% - #{$input-margin-bottom});
    z-index: 10;
  }

  .tag-selector__dropdown-list {
    @extend %vf-list;
    margin-bottom: 0;
    max-height: $dropdown-max-height;
    overflow: auto;
  }

  .tag-selector__dropdown-header {
    margin-bottom: 0;
    padding: $spv--x-small $sph--small;
    position: relative;

    &::after {
      border-bottom: 1px solid $color-mid-light;
      bottom: 0;
      content: "";
      height: 1px;
      left: 0;
      position: absolute;
      right: 0;
    }
  }

  .tag-selector__dropdown-item {
    @extend %vf-list-item;
    @include vf-list-item-divided;
  }

  .tag-selector__dropdown-item-description {
    @extend %small-text;
    color: $color-mid-dark;
  }

  .tag-selector__dropdown-button {
    border: 0;
    margin-bottom: 0;
    padding-left: $sph--small;
    padding-right: $sph--small;
    text-align: left;
    width: 100%;
  }

  .tag-selector__selected-list {
    background-color: $colors--light-theme--background-inputs;
    border-bottom: 0;
    margin: 0;
    padding: $spv--x-small $sph--small;
  }

  .is-error .tag-selector__selected-list {
    border-color: $color-negative;
  }

  .tag-selector__selected-item {
    @include vf-inline-list-item;
  }

  .tag-selector__selected-button {
    padding: 0 $sph--small 0 0 !important;
  }
}
