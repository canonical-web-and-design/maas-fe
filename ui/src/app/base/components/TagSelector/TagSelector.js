import { Button, Input, Label } from "@canonical/react-components";
import classNames from "classnames";
import React, { useEffect, useRef, useState } from "react";

/**
 * Highlights a portion of given text that matches substring.
 * @param {String} text - Text to search for substring.
 * @param {String} match - Substring to highlight.
 * @returns {Object} JSX with emphasised text.
 */
const highlightMatch = (text, match) => {
  const textArray = text.split(match);
  return (
    <span>
      {textArray.map((item, i) => (
        <span key={`${item}${i}`}>
          {item}
          {i !== textArray.length - 1 && <em>{match}</em>}
        </span>
      ))}
    </span>
  );
};

const sanitiseFilter = (filterText) => filterText.replace(/ /g, "-");

const generateDropdownItems = ({ filter, selectedTags, tags, updateTags }) => {
  const dropdownItems = [];
  if (
    filter &&
    !tags.some((tag) => tag === filter) &&
    !selectedTags.includes(filter)
  ) {
    // Insert an extra item for creating a new tag if filter is present and it
    // is neither an existing tag nor already in the list of selected tags.
    const newTagItem = (
      <li className="tag-selector__dropdown-item" key={filter}>
        <Button
          appearance="base"
          className="tag-selector__dropdown-button"
          data-test="new-tag"
          onClick={() => {
            updateTags([...selectedTags, sanitiseFilter(filter)]);
          }}
          type="button"
        >
          <em>Create tag "{filter}"</em>
        </Button>
      </li>
    );
    dropdownItems.push(newTagItem);
  }

  const existingTagItems = tags
    .filter((tag) => !selectedTags.includes(tag) && tag.includes(filter))
    .map((tag) => (
      <li className="tag-selector__dropdown-item" key={tag}>
        <Button
          appearance="base"
          className="tag-selector__dropdown-button"
          data-test="existing-tag"
          onClick={() => {
            updateTags([...selectedTags, tag]);
          }}
          type="button"
        >
          {filter ? highlightMatch(tag, filter) : tag}
        </Button>
      </li>
    ));
  dropdownItems.push(existingTagItems);
  return dropdownItems;
};

const generateSelectedItems = ({ selectedTags, updateTags }) =>
  selectedTags.map((tag) => (
    <li className="tag-selector__selected-item" key={tag}>
      <Button
        appearance="base"
        className="tag-selector__selected-button"
        data-test="selected-tag"
        dense
        hasIcon
        onClick={() =>
          updateTags(
            selectedTags.filter((item) => item !== tag),
            false
          )
        }
      >
        <span>{tag}</span>
        <i className="p-icon--close" />
      </Button>
    </li>
  ));

export const TagSelector = ({
  initialSelected = [],
  label,
  onTagsUpdate,
  placeholder = "Tags",
  required,
  tags,
}) => {
  const wrapperRef = useRef(null);

  const [dropdownOpen, setDropdownOpen] = useState(false);
  const [selectedTags, setSelectedTags] = useState(initialSelected);
  const [filter, setFilter] = useState("");

  const updateTags = (newSelectedTags, clearFilter = true) => {
    const sortedTags = newSelectedTags.sort((a, b) => a.localeCompare(b));
    setSelectedTags(sortedTags);
    onTagsUpdate && onTagsUpdate(sortedTags);
    clearFilter && setFilter("");
  };

  const handleClickOutside = (e) => {
    if (
      wrapperRef.current &&
      !wrapperRef.current.contains(e.target) &&
      !e.target.className.includes("tag-selector")
    ) {
      setDropdownOpen(false);
    }
  };

  useEffect(() => {
    document.addEventListener("click", handleClickOutside, false);
    return () => {
      document.removeEventListener("click", handleClickOutside, false);
    };
  }, []);

  return (
    <div className="tag-selector" ref={wrapperRef}>
      <Label onClick={() => setDropdownOpen(true)}>{label}</Label>
      {selectedTags.length > 0 && (
        <ul className="tag-selector__selected-list">
          {generateSelectedItems({ selectedTags, updateTags })}
        </ul>
      )}
      <Input
        className={classNames("tag-selector__input", {
          "tags-selected": selectedTags.length > 0,
        })}
        onChange={(e) => setFilter(e.target.value)}
        onFocus={() => setDropdownOpen(true)}
        onKeyPress={(e) => {
          if (e.key === "Enter") {
            e.preventDefault();
            updateTags([...selectedTags, sanitiseFilter(filter)]);
          }
        }}
        placeholder={placeholder}
        required={required}
        type="text"
        value={filter}
      />
      {dropdownOpen && (
        <div className="tag-selector__dropdown">
          <ul className="tag-selector__dropdown-list">
            {generateDropdownItems({
              filter,
              selectedTags,
              tags,
              updateTags,
            })}
          </ul>
        </div>
      )}
    </div>
  );
};

export default TagSelector;
