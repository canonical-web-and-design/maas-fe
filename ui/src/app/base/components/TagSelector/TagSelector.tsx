import React, { useEffect, useRef, useState } from "react";

import { Button, Input } from "@canonical/react-components";
import Field from "@canonical/react-components/dist/components/Field";
import classNames from "classnames";

export type Tag = {
  id: number;
  name: string;
  displayName?: string;
  description?: string;
};

/**
 * Highlights a portion of given text that matches substring.
 * @param text - Text to search for substring.
 * @param match - Substring to highlight.
 * @returns JSX with emphasised text.
 */
const highlightMatch = (text: string, match: string): JSX.Element => {
  const textArray = text.split(match);
  return (
    <span>
      {textArray.map((item, i) => (
        <span key={`${item}${i}`}>
          {item}
          {i !== textArray.length - 1 && <em>{match}</em>}
        </span>
      ))}
    </span>
  );
};

const sanitiseFilter = (filterText: string) => ({
  name: filterText.replace(/ /g, "-"),
});

const generateDropdownItems = ({
  allowNewTags,
  filter,
  selectedTags,
  tags,
  updateTags,
}): JSX.Element[] => {
  const dropdownItems = [];
  if (
    allowNewTags &&
    filter &&
    !tags.some((tag) => (tag.displayName || tag.name) === filter) &&
    !selectedTags.some((tag) => (tag.displayName || tag.name) === filter)
  ) {
    // Insert an extra item for creating a new tag if allowed and filter is not
    // an already existing tag
    const newTagItem = (
      <li className="tag-selector__dropdown-item" key={filter}>
        <Button
          appearance="base"
          className="tag-selector__dropdown-button"
          data-test="new-tag"
          onClick={() => {
            updateTags([...selectedTags, sanitiseFilter(filter)]);
          }}
          type="button"
        >
          <em>Create tag "{filter}"</em>
        </Button>
      </li>
    );
    dropdownItems.push(newTagItem);
  }

  const existingTagItems = tags
    .filter(
      (tag) =>
        (tag.displayName || tag.name).includes(filter) &&
        !selectedTags.some((selectedTag) => selectedTag.name === tag.name)
    )
    .map((tag) => (
      <li className="tag-selector__dropdown-item" key={tag.name}>
        <Button
          appearance="base"
          className="tag-selector__dropdown-button"
          data-test="existing-tag"
          onClick={() => {
            updateTags([...selectedTags, tag]);
          }}
          type="button"
        >
          {filter
            ? highlightMatch(tag.displayName || tag.name, filter)
            : tag.displayName || tag.name}
          {tag.description && (
            <div className="tag-selector__dropdown-item-description">
              {tag.description}
            </div>
          )}
        </Button>
      </li>
    ));

  return dropdownItems.concat(existingTagItems);
};

const generateSelectedItems = (
  selectedTags: Tag[],
  updateTags,
  disabledTags: Tag[]
) =>
  selectedTags.map((tag) => {
    const isDisabled = disabledTags?.some(
      (disabledTag) => disabledTag.id === tag.id
    );

    return (
      <li className="tag-selector__selected-item" key={tag.name}>
        <Button
          appearance="base"
          className="tag-selector__selected-button"
          data-test="selected-tag"
          disabled={isDisabled}
          dense
          hasIcon
          onClick={() =>
            updateTags(
              selectedTags.filter((item) => item !== tag),
              false
            )
          }
          type="button"
        >
          <span>{tag.name}</span>
          {!isDisabled && <i className="p-icon--close" />}
        </Button>
      </li>
    );
  });

type Props = {
  allowNewTags?: boolean;
  disabled?: boolean;
  error?: string;
  help?: string;
  initialSelected?: Tag[];
  label?: string;
  onTagsUpdate?: (tags: Tag[]) => void;
  placeholder: string;
  required?: boolean;
  tags: Tag[];
  disabledTags?: Tag[];
};

export const TagSelector = ({
  allowNewTags = false,
  disabled,
  error,
  help,
  initialSelected = [],
  label,
  onTagsUpdate,
  placeholder = "Tags",
  required = false,
  tags = [],
  disabledTags = [],
}: Props): JSX.Element => {
  const wrapperRef = useRef(null);

  const [dropdownOpen, setDropdownOpen] = useState(false);
  const [selectedTags, setSelectedTags] = useState(initialSelected);
  const [filter, setFilter] = useState("");

  const updateTags = (newSelectedTags, clearFilter = true) => {
    const sortedTags = newSelectedTags.sort((a, b) =>
      a.name.localeCompare(b.name)
    );
    setSelectedTags(sortedTags);
    onTagsUpdate && onTagsUpdate(sortedTags);
    clearFilter && setFilter("");
  };

  const handleClickOutside = (e) => {
    if (
      wrapperRef.current &&
      !wrapperRef.current.contains(e.target) &&
      !e.target.className.includes("tag-selector")
    ) {
      setDropdownOpen(false);
    }
  };

  useEffect(() => {
    document.addEventListener("click", handleClickOutside, false);
    return () => {
      document.removeEventListener("click", handleClickOutside, false);
    };
  }, []);

  const dropdownItems = generateDropdownItems({
    allowNewTags,
    filter,
    selectedTags,
    tags,
    updateTags,
  });

  return (
    <Field
      error={error}
      help={help}
      label={
        label ? (
          <span onClick={() => setDropdownOpen(true)} ref={wrapperRef}>
            {label}
          </span>
        ) : undefined
      }
    >
      <div className="tag-selector">
        {selectedTags.length > 0 && (
          <ul className="tag-selector__selected-list">
            {generateSelectedItems(selectedTags, updateTags, disabledTags)}
          </ul>
        )}
        <Input
          className={classNames("tag-selector__input", {
            "tags-selected": selectedTags.length > 0,
          })}
          disabled={disabled}
          onChange={(e) => setFilter(e.target.value)}
          onFocus={() => setDropdownOpen(true)}
          onKeyPress={(e) => {
            if (e.key === "Enter") {
              e.preventDefault();
              if (allowNewTags) {
                updateTags([...selectedTags, sanitiseFilter(filter)]);
              }
            }
          }}
          placeholder={selectedTags.length !== tags.length ? placeholder : ""}
          required={required}
          type="text"
          value={filter}
        />
        {dropdownOpen && dropdownItems.length >= 1 && (
          <div className="tag-selector__dropdown">
            <ul className="tag-selector__dropdown-list">{dropdownItems}</ul>
          </div>
        )}
      </div>
    </Field>
  );
};

export default TagSelector;
