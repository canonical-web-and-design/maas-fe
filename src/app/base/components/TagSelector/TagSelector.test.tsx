import { mount, shallow } from "enzyme";

import type { Tag } from "./TagSelector";
import TagSelector from "./TagSelector";

describe("TagSelector", () => {
  let tags: Tag[];
  beforeEach(() => {
    tags = [
      { displayName: "tag one", name: "tag1" },
      { displayName: "tag two", name: "tag2" },
    ];
  });

  // snapshot tests
  it("renders and matches the snapshot when closed", () => {
    const component = shallow(
      <TagSelector
        label="Tags"
        onTagsUpdate={jest.fn()}
        placeholder="Select or create tags"
        tags={tags}
      />
    );
    expect(component).toMatchSnapshot();
  });

  it("renders and matches the snapshot when opened", () => {
    const component = mount(
      <TagSelector
        label="Tags"
        onTagsUpdate={jest.fn()}
        placeholder="Select or create tags"
        tags={tags}
      />
    );
    component.find("Label").first().simulate("click");
    expect(component).toMatchSnapshot();
  });

  it("renders and matches the snapshot with tag descriptions", () => {
    const component = mount(
      <TagSelector
        label="Tags"
        onTagsUpdate={jest.fn()}
        placeholder="Select or create tags"
        tags={[
          { ...tags[0], description: "description one" },
          { ...tags[1], description: "description two" },
        ]}
      />
    );
    component.find("Label").first().simulate("click");
    expect(component).toMatchSnapshot();
  });

  // unit tests
  it("can have some tags preselected", () => {
    const component = shallow(
      <TagSelector
        initialSelected={[tags[0]]}
        label="Tags"
        onTagsUpdate={jest.fn()}
        placeholder="Select or create tags"
        tags={tags}
      />
    );
    expect(
      component.find('[data-testid="selected-tag"] span').at(0).text()
    ).toBe("tag1");
  });

  it("opens the dropdown when input is focused", () => {
    const component = shallow(
      <TagSelector
        label="Tags"
        onTagsUpdate={jest.fn()}
        placeholder="Select or create tags"
        tags={tags}
      />
    );
    expect(component.find(".tag-selector__dropdown").exists()).toBe(false);
    component.find(".tag-selector__input").simulate("focus");
    expect(component.find(".tag-selector__dropdown").exists()).toBe(true);
  });

  it("can select existing tags from dropdown", () => {
    const component = shallow(
      <TagSelector
        label="Tags"
        onTagsUpdate={jest.fn()}
        placeholder="Select or create tags"
        tags={tags}
      />
    );
    component.find(".tag-selector__input").simulate("focus");
    component.find('[data-testid="existing-tag"]').at(0).simulate("click");
    expect(
      component.find('[data-testid="selected-tag"] span').at(0).text()
    ).toBe("tag1");
  });

  it("can hide the tags that have been selected", () => {
    const component = shallow(
      <TagSelector
        label="Tags"
        onTagsUpdate={jest.fn()}
        placeholder="Select or create tags"
        showSelectedTags={false}
        tags={tags}
      />
    );
    component.find(".tag-selector__input").simulate("focus");
    component.find('[data-testid="existing-tag"]').at(0).simulate("click");
    expect(component.find('[data-testid="selected-tag"] span').exists()).toBe(
      false
    );
    expect(component.find(".tag-selector__selected-list").exists()).toBe(false);
  });

  it("can remove tags that have been selected", () => {
    const component = shallow(
      <TagSelector
        initialSelected={tags}
        label="Tags"
        onTagsUpdate={jest.fn()}
        placeholder="Select or create tags"
        tags={tags}
      />
    );
    expect(component.find('[data-testid="selected-tag"]').length).toBe(2);
    component.find(".tag-selector__input").simulate("focus");
    component.find('[data-testid="selected-tag"]').at(0).simulate("click");
    expect(component.find('[data-testid="selected-tag"]').length).toBe(1);
    expect(
      component.find('[data-testid="selected-tag"] span').at(0).text()
    ).toBe("tag2");
  });

  it("can create and select a new tag", () => {
    const component = shallow(
      <TagSelector
        allowNewTags
        label="Tags"
        onTagsUpdate={jest.fn()}
        placeholder="Select or create tags"
        tags={tags}
      />
    );
    component.find(".tag-selector__input").simulate("focus");
    component
      .find(".tag-selector__input")
      .simulate("change", { target: { value: "new-tag" } });
    component.find('[data-testid="new-tag"]').simulate("click");
    expect(
      component.find('[data-testid="selected-tag"] span').at(0).text()
    ).toBe("new-tag");
  });

  it("can call a provide function to create a new tag", () => {
    const onAddNewTag = jest.fn();
    const component = shallow(
      <TagSelector
        allowNewTags
        label="Tags"
        onAddNewTag={onAddNewTag}
        onTagsUpdate={jest.fn()}
        placeholder="Select or create tags"
        tags={tags}
      />
    );
    component.find(".tag-selector__input").simulate("focus");
    component
      .find(".tag-selector__input")
      .simulate("change", { target: { value: "new-tag" } });
    component.find('[data-testid="new-tag"]').simulate("click");
    expect(onAddNewTag).toHaveBeenCalledWith("new-tag");
    // The input should get cleared.
    expect(component.find(".tag-selector__input").prop("value")).toBe("");
  });

  it("sanitises text when creating new tag", () => {
    const component = shallow(
      <TagSelector
        allowNewTags
        label="Tags"
        onTagsUpdate={jest.fn()}
        placeholder="Select or create tags"
        tags={tags}
      />
    );
    component.find(".tag-selector__input").simulate("focus");
    component
      .find(".tag-selector__input")
      .simulate("change", { target: { value: "tag with spaces" } });
    component.find('[data-testid="new-tag"]').simulate("click");
    expect(
      component.find('[data-testid="selected-tag"] span').at(0).text()
    ).toBe("tag-with-spaces");
  });

  it("can filter tag list", () => {
    const component = shallow(
      <TagSelector
        label="Tags"
        onTagsUpdate={jest.fn()}
        placeholder="Select or create tags"
        tags={[...tags, { displayName: "other", name: "other" }]}
      />
    );
    component.find(".tag-selector__input").simulate("focus");
    expect(component.find('[data-testid="existing-tag"]').length).toBe(3);
    component
      .find(".tag-selector__input")
      .simulate("change", { target: { value: "tag" } });
    expect(component.find('[data-testid="existing-tag"]').length).toBe(2);
  });

  it("can highlight what matches the filter in existing tags", () => {
    const component = shallow(
      <TagSelector
        label="Tags"
        onTagsUpdate={jest.fn()}
        placeholder="Select or create tags"
        tags={[
          { displayName: "there", name: "there" },
          { displayName: "other", name: "other" },
        ]}
      />
    );
    component.find(".tag-selector__input").simulate("focus");
    component
      .find(".tag-selector__input")
      .simulate("change", { target: { value: "the" } });
    expect(
      component.find('[data-testid="existing-tag"] > span').at(0).text()
    ).toBe("there");
    expect(
      component.find('[data-testid="existing-tag"] > span strong').at(0).text()
    ).toBe("the");
    expect(
      component.find('[data-testid="existing-tag"] > span').at(1).text()
    ).toBe("other");
    expect(
      component.find('[data-testid="existing-tag"] > span strong').at(1).text()
    ).toBe("the");
  });

  it("can disable tags", () => {
    const tags = [
      { id: 1, name: "enabledTag" },
      { id: 2, name: "disabledTag" },
    ];
    const component = shallow(
      <TagSelector
        disabledTags={[{ id: 2, name: "disabledTag" }]}
        initialSelected={tags}
        tags={tags}
      />
    );

    expect(
      component.find('[data-testid="selected-tag"]').at(0).prop("disabled")
    ).toBe(false);
    expect(
      component.find('[data-testid="selected-tag"]').at(1).prop("disabled")
    ).toBe(true);
  });

  it("can display a dropdown header", () => {
    const component = shallow(
      <TagSelector
        header={<span data-testid="dropdown-header"></span>}
        label="Tags"
        onTagsUpdate={jest.fn()}
        placeholder="Select or create tags"
        showSelectedTags={false}
        tags={tags}
      />
    );
    component.find(".tag-selector__input").simulate("focus");
    const header = component.find(".tag-selector__dropdown");
    expect(header.exists()).toBe(true);
    expect(header.find("[data-testid='dropdown-header']").exists()).toBe(true);
  });

  it("can customise the dropdown items", () => {
    const component = shallow(
      <TagSelector
        generateDropdownEntry={() => <span data-testid="dropdown-item"></span>}
        label="Tags"
        onTagsUpdate={jest.fn()}
        placeholder="Select or create tags"
        showSelectedTags={false}
        tags={tags}
      />
    );
    component.find(".tag-selector__input").simulate("focus");
    expect(
      component
        .find(".tag-selector__dropdown-button [data-testid='dropdown-item']")
        .exists()
    ).toBe(true);
  });

  it("can use an external list of selected tags", () => {
    const component = shallow(
      <TagSelector
        externalSelectedTags={[tags[0]]}
        label="Tags"
        onTagsUpdate={jest.fn()}
        tags={tags}
      />
    );
    expect(
      component.find('[data-testid="selected-tag"] span').at(0).text()
    ).toBe("tag1");
  });

  it("handles selecting external tags", () => {
    const onTagsUpdate = jest.fn();
    const component = shallow(
      <TagSelector
        externalSelectedTags={[tags[0]]}
        label="Tags"
        onTagsUpdate={onTagsUpdate}
        tags={tags}
      />
    );
    component.find(".tag-selector__input").simulate("focus");
    component.find('[data-testid="existing-tag"]').at(0).simulate("click");
    expect(onTagsUpdate).toHaveBeenCalledWith([tags[0], tags[1]]);
  });
});
