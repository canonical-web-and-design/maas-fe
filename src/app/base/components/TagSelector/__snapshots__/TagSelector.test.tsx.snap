// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TagSelector renders and matches the snapshot when closed 1`] = `
<div>
  <Field
    label={
      <span
        onClick={[Function]}
      >
        Tags
      </span>
    }
  >
    <div
      className="tag-selector"
    >
      <Input
        aria-haspopup="listbox"
        aria-label="Tags"
        className="tag-selector__input"
        onChange={[Function]}
        onFocus={[Function]}
        onKeyPress={[Function]}
        placeholder="Select or create tags"
        required={false}
        type="text"
        value=""
      />
    </div>
  </Field>
</div>
`;

exports[`TagSelector renders and matches the snapshot when opened 1`] = `
<TagSelector
  label="Tags"
  onTagsUpdate={[MockFunction]}
  placeholder="Select or create tags"
  tags={
    Array [
      Object {
        "displayName": "tag one",
        "name": "tag1",
      },
      Object {
        "displayName": "tag two",
        "name": "tag2",
      },
    ]
  }
>
  <div>
    <Field
      label={
        <span
          onClick={[Function]}
        >
          Tags
        </span>
      }
    >
      <div
        className="p-form__group p-form-validation"
      >
        <Label>
          <label
            className="p-form__label"
          >
            <span
              onClick={[Function]}
            >
              Tags
            </span>
          </label>
        </Label>
        <div
          className="p-form__control u-clearfix"
        >
          <div
            className="tag-selector"
          >
            <Input
              aria-haspopup="listbox"
              aria-label="Tags"
              className="tag-selector__input"
              onChange={[Function]}
              onFocus={[Function]}
              onKeyPress={[Function]}
              placeholder="Select or create tags"
              required={false}
              type="text"
              value=""
            >
              <Field
                helpId="mock-nanoid-2"
                isTickElement={false}
                required={false}
                validationId="mock-nanoid-1"
              >
                <div
                  className="p-form__group p-form-validation"
                >
                  <div
                    className="p-form__control u-clearfix"
                  >
                    <input
                      aria-describedby=""
                      aria-errormessage={null}
                      aria-haspopup="listbox"
                      aria-invalid={false}
                      aria-label="Tags"
                      className="p-form-validation__input tag-selector__input"
                      onChange={[Function]}
                      onFocus={[Function]}
                      onKeyPress={[Function]}
                      placeholder="Select or create tags"
                      required={false}
                      type="text"
                      value=""
                    />
                  </div>
                </div>
              </Field>
            </Input>
          </div>
        </div>
      </div>
    </Field>
  </div>
</TagSelector>
`;

exports[`TagSelector renders and matches the snapshot with tag descriptions 1`] = `
<TagSelector
  label="Tags"
  onTagsUpdate={[MockFunction]}
  placeholder="Select or create tags"
  tags={
    Array [
      Object {
        "description": "description one",
        "displayName": "tag one",
        "name": "tag1",
      },
      Object {
        "description": "description two",
        "displayName": "tag two",
        "name": "tag2",
      },
    ]
  }
>
  <div>
    <Field
      label={
        <span
          onClick={[Function]}
        >
          Tags
        </span>
      }
    >
      <div
        className="p-form__group p-form-validation"
      >
        <Label>
          <label
            className="p-form__label"
          >
            <span
              onClick={[Function]}
            >
              Tags
            </span>
          </label>
        </Label>
        <div
          className="p-form__control u-clearfix"
        >
          <div
            className="tag-selector"
          >
            <Input
              aria-haspopup="listbox"
              aria-label="Tags"
              className="tag-selector__input"
              onChange={[Function]}
              onFocus={[Function]}
              onKeyPress={[Function]}
              placeholder="Select or create tags"
              required={false}
              type="text"
              value=""
            >
              <Field
                helpId="mock-nanoid-2"
                isTickElement={false}
                required={false}
                validationId="mock-nanoid-1"
              >
                <div
                  className="p-form__group p-form-validation"
                >
                  <div
                    className="p-form__control u-clearfix"
                  >
                    <input
                      aria-describedby=""
                      aria-errormessage={null}
                      aria-haspopup="listbox"
                      aria-invalid={false}
                      aria-label="Tags"
                      className="p-form-validation__input tag-selector__input"
                      onChange={[Function]}
                      onFocus={[Function]}
                      onKeyPress={[Function]}
                      placeholder="Select or create tags"
                      required={false}
                      type="text"
                      value=""
                    />
                  </div>
                </div>
              </Field>
            </Input>
          </div>
        </div>
      </div>
    </Field>
  </div>
</TagSelector>
`;
